src/Bicep.LangServer/Handlers/BicepDefinitionHandler.cs

```diff
@@ -2,61 +2,242 @@
         {
-            var result = this.symbolResolver.ResolveSymbol(request.TextDocument.Uri, request.Position);
-            if (result == null)
+            var context = this.compilationManager.GetCompilation(request.TextDocument.Uri);
+            if (context is null)
             {
                 return Task.FromResult(new LocationOrLocationLinks());
             }
+            var result = this.symbolResolver.ResolveSymbol(request.TextDocument.Uri, request.Position);
+            // No parent Symbol: ad hoc syntax matching
+            if (result is null)
+            {
+                return HandleUnboundSymbolLocationAsync(request, context);
+            }
+            // Declared symbols: go to definition
+            else if (result.Symbol is DeclaredSymbol declaration)
+            {
+                return DeclaredDefinitionLocationAsync(request, result, declaration);
+            }
+            // Object property: currently only used for module param goto
+            else if (result.Origin is ObjectPropertySyntax objectPropertySyntax)
+            {
+                return GetObjectPropertyLocationAsync(request, result, context);
+            }
+            // Used for module (name), variable or resource property access
+            else if (result.Symbol is PropertySymbol)
+            {
+                return GetPropertyLocationAsync(request, result, context);
+            }
+            return Task.FromResult(new LocationOrLocationLinks()); 
+        }
+
+        protected override DefinitionRegistrationOptions CreateRegistrationOptions(DefinitionCapability capability, ClientCapabilities clientCapabilities) => new()
+        {
+            DocumentSelector = DocumentSelectorFactory.Create()
+        };
 
-            if (result.Symbol is PropertySymbol)
+        private Task<LocationOrLocationLinks> HandleUnboundSymbolLocationAsync(DefinitionParams request, CompilationContext context)
+        {
+            // Currently we only definition handler for a non symbol bound to implement module path goto.
+            // try to resolve module path syntax from given offset using tail matching.
+            int offset = PositionHelper.GetOffset(context.LineStarts, request.Position);
+            var matchingNodes = SyntaxMatcher.FindNodesMatchingOffset(context.Compilation.SourceFileGrouping.EntryPoint.ProgramSyntax, offset);
+            if (SyntaxMatcher.IsTailMatch<ModuleDeclarationSyntax, StringSyntax, Token>(
+                matchingNodes,
+                (moduleSyntax, stringSyntax, token) => moduleSyntax.Path == stringSyntax && token.Type == TokenType.StringComplete)
+                && matchingNodes[^3] is ModuleDeclarationSyntax moduleDeclarationSyntax
+                && matchingNodes[^2] is StringSyntax stringToken
+                && context.Compilation.SourceFileGrouping.LookUpModuleSourceFile(moduleDeclarationSyntax) is BicepFile bicepFile)
```

Comment:
>BicepFile [](http://example.com/codeflow?start=109&length=9)

Modules support referencing JSON files as well. What's the plan for those?

Reply:
(We can defer this to another PR to keep this one small.)

Reply:
Good catch, was able to change this to support goto for json files. As for outputs and parameters goto, worth triaging.